### ISSUE TITLE
Attractive Spruce Beetle - Forge signatures on different proposals on castVoteBySig function

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-01-olympus-on-chain-governance-judging#24]
Attractive Spruce Beetle

high

# Forge signatures on different proposals on castVoteBySig function

## Summary
Reuse of signature data. The signature data is not unique on each proposal ID, this  means that the same sig data can be used on different proposals.

## Vulnerability Detail

The` castVoteBySig` function reuses the same signature data (digest) for all votes, regardless of the proposal ID. This could allow an attacker to forge signatures for votes on different proposals by reusing the same signature data and changing the support value.

### GovernorBravoDelegate.castVoteBySig
```solidity
 function castVoteBySig(
        uint256 proposalId,
        uint8 support,
        uint8 v,
        bytes32 r,
        bytes32 s
    ) external {
        bytes32 domainSeparator = keccak256(
            abi.encode(DOMAIN_TYPEHASH, keccak256(bytes(name)), getChainIdInternal(), address(this))
        );
        bytes32 structHash = keccak256(abi.encode(BALLOT_TYPEHASH, proposalId, support));
        bytes32 digest = keccak256(abi.encodePacked("\x19\x01", domainSeparator, structHash));  //[x]
        address signatory = ecrecover(digest, v, r, s);                              // [x]

        if (signatory == address(0)) revert GovernorBravo_InvalidSignature();
        emit VoteCast(
            signatory,
            proposalId,
            support,
            castVoteInternal(signatory, proposalId, support),
            ""
        );
    }
```
## Impact
Attacker can vote on different proposals on behalf of a user.

## Code Snippet
https://github.com/sherlock-audit/2024-01-olympus-on-chain-governance/blob/main/bophades/src/external/governance/GovernorBravoDelegate.sol#L403

## Tool used

Manual Review

## Recommendation
The `castVoteBySig` function should protect against replay attacks. It should use a nonce or other unique identifier to prevent an attacker from replaying a valid signature to cast multiple votes on the same proposal.

**Comment by 0xLienid** [comment]
this is not true. the proposalId is used in the structHash which is used in the digest
